Command to train list_models

ravih18 opened the pull request:
Added a new command to train command group to get the list of available models in clinicadl and their detailed architecture.

mdiazmel commented:
Thx @ravih18 ,

Could you please add a word about this new interface in the official documentation?

14thibea reviewed (approved):
Hello @ravih18 !

Thanks for the PR! It seems that it is not working for me right now.
When I type `clinicadl train list_models` I only see the list of options (`--architecture` and `--input_size`) and not the list of architectures available.

Moreover if I choose a specific architecture, `clinicadl train list_models Conv5_FC3` throws an error:
```
Traceback (most recent call last):
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/bin/clinicadl", line 5, in <module>
    cli()
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/elina.thibeausutre/miniconda3/envs/clinicadl_fork-RH/lib/python3.7/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/elina.thibeausutre/Documents/code/clinicadl_fork-RH/clinicadl/train/list_models_cli.py", line 29, in cli
    get_model_list(architecture, input_size)
  File "/Users/elina.thibeausutre/Documents/code/clinicadl_fork-RH/clinicadl/train/train_utils.py", line 97, in get_model_list
    model = model_class(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'use_cpu'
```
Actually the argument `use_cpu` was renamed `gpu` this must be corrected in the PR.

14thibea commented on clinicadl/train/train_utils.py:
```diff
@@ -67,3 +67,32 @@ def get_user_dict(config_file: str, task: str) -> Dict[str, Any]:
+        args.remove("self")
+        kwargs = dict()
+        kwargs["input_size"] = (1, input_size[0], input_size[1])
+        kwargs["use_cpu"] = True
```
```suggestion
        kwargs["gpu"] = False
```

14thibea commented on clinicadl/train/train_utils.py:
```diff
@@ -67,3 +67,32 @@ def get_user_dict(config_file: str, task: str) -> Dict[str, Any]:
+    if not architecture:
+        model_list = getmembers(network_package, isclass)
+        for model in model_list:
+            print(model[0])
```
Maybe this would work using `click.echo` instead of `print`?

ravih18 commented:
Hello, than you for your review @14thibea .
`use_cpu` option has been replaced by `gpu`.
`print` has been replaced by `click.echo`.
I added a few sentences that can maybe make the printed message clearer.
Finally I removed the `no_args_is_help` which was having a wrong behavior. Indeed if this command line is used without arguments (and indeed it doesn't take arguments) it is supposed to show the list of all models that the network module can load. This is now the case !

14thibea commented on clinicadl/train/train_utils.py:
```diff
@@ -67,3 +67,40 @@ def get_user_dict(config_file: str, task: str) -> Dict[str, Any]:
+    import clinicadl.utils.network as network_package
+
+    if not architecture:
+        echo("The list of currently available model is:")
```
```suggestion
        echo("The list of currently available models is:")
```

14thibea commented on clinicadl/train/train_utils.py:
```diff
@@ -67,3 +67,40 @@ def get_user_dict(config_file: str, task: str) -> Dict[str, Any]:
+        )
+        args.remove("self")
+        kwargs = dict()
+        kwargs["input_size"] = (1, input_size[0], input_size[1])
```
Actually this won't work for 3D architectures, and we might want to know the size of the FC layers for Conv5_FC3 in a 3D case.
It could be better not to be dependent of the dimenson, and just let the users set the input_size they want
```suggestion
        kwargs["input_size"] = input_size
```

14thibea commented on clinicadl/train/list_models_cli.py:
```diff
@@ -0,0 +1,33 @@
+    "-in",
+    "--input_size",
+    type=int,
+    multiple=2,
+    default=(128, 128),
+    show_default=True,
+    help="Size of the input image.",
```
In order to work with the 3D case, it could be better to use `nargs=True` and have a default value including the number of channels (which may be 1 or 3 depending on the slice setting).
```suggestion
    nargs=True,
    default=(1, 128, 128),
    show_default=True,
    help="Size of the input image.",
```

14thibea commented on clinicadl/train/train_utils.py:
```diff
@@ -67,3 +67,40 @@ def get_user_dict(config_file: str, task: str) -> Dict[str, Any]:
+
+        model = model_class(**kwargs)
+        secho(f"Information for {architecture} network", bold=True)
+        echo("Model layers:")
```
It could be nice to print the input size used to generate the model as the size of FC layers depends on the size of the input.
```suggestion
        echo(f"Input size: {input_size}")
        echo("Model layers:")
```

14thibea commented on docs/Train/Details.md:
```diff
@@ -115,6 +115,16 @@ the current state of the model is saved in the corresponding best model folder.
+There are a few options to get more details:
+
+- `--architecture`: you can specify there the name a model and it will display the detail architecture corresponding to the model.
+- `--input_size`: to specify the size of you input (if the model is 2D).
```
This should work even if the model is 3D
```suggestion
- `--input_size`: is the size of the input of the model. For example if the input size is 1@169x208x179, the option `--input_size 1 169 208 179` should be added. Default value: 1@128x128.
```

14thibea commented on docs/Train/Details.md:
```diff
@@ -115,6 +115,16 @@ the current state of the model is saved in the corresponding best model folder.
+will display the list of available models that can be used.
+There are a few options to get more details:
+
+- `--architecture`: you can specify there the name a model and it will display the detail architecture corresponding to the model.
```
```suggestion
- `--architecture`: is the name of the model whose architecture will be displayed.
```

14thibea reviewed (commented):
It's working for me now!
I just suggested some changes to make everything work with 3D input sizes, as changing the size of the input may change the architecture.

14thibea commented on docs/Train/Details.md:
```diff
@@ -115,6 +115,18 @@ the current state of the model is saved in the corresponding best model folder.
+- `--architecture`: is the name of the model whose architecture will be displayed.
+- `--input_size`: is the size of the input of the model with shape C@HxW if the image is 2D or C@DxHxW if the image is 3D. 
+  For example if the input size is 1@169x208x179, the option `--input_size 1@169x208x179` should be added.
+  Default value: `128x128`.
```
```suggestion
  Default value: `1@128x128`.
```

14thibea reviewed (approved):
Good for me!
Just a remaining typo in the documentation

mdiazmel commented:
Nice, LGTM!
